### ISSUE TITLE
Request for restoring the deprecated AEAD cipher "XChaCha20-Poly1305-IETF"

### ISSUE THREAD

**Issue by luminouser** [shadowsocks/shadowsocks-android#2663]
Please read contributing guidelines. Thanks.

**Is your feature request related to a problem? Please describe.**
It is thought that stream ciphers are unsafe for shadowsocks. However, XChaCha20-Poly1305-IETF is an AEAD construction, not just a stream cipher. 
There is no clear evidence that AEAD constructions with stream ciphers are not safe. And stream ciphers are usually quicker than block ciphers. Furthermore, XChaCha20-Poly1305-IETF is the officially recommended cipher in libsodium documentation. [read this](https://doc.libsodium.org/secret-key_cryptography/aead#tl-dr-which-one-should-i-use) 

**Describe the solution you'd like**
Restoring the "XChaCha20-Poly1305-IETF" encryption scheme

**Describe alternatives you've considered**
None.

**Additional context**
None.

**Comment by gengen1988** [comment]
TLDR: It's impossible for now.

Long story:

In previous versions, this project using shadowsocks-libev to provide shadowsocks client implements. It relied on libsodium which has many experimental ciphers including XChaCha20-Poly1305-IETF. In mid 2020, shadowsocks-android switch to shadowsocks-rust as backend. It using crpyto2 as cipher implements. The crypto2 has no XChaCha20-Poly1305-IETF for some reason. That is why you can't use XChaCha20-Poly1305-IETF anymore.

If you wish to switch back old school libsodium, you may fork from v5.0.5

References:

https://github.com/shadowsocks/shadowsocks-rust/pull/319 (Chinese)
https://github.com/shadowsocks/shadowsocks-libev
https://github.com/shadowsocks/shadowsocks-rust
https://github.com/shadowsocks/crypto2
https://github.com/jedisct1/libsodium

**Comment by Mygod** [comment]
@zonyitoo @madeye Transfer this issue to shadowsocks-rust, or close?

**Comment by zonyitoo** [comment]
It could be transferred to shadowsocks-rust.

A short version of why crypto2 won't support xchacha20-poly1305-ietf:

xchacha20-poly1305-ietf is invented by libsodium and it is the only implementation the the whole world. They (libsodium) tried to make this algorithm become [RFC standard](https://tools.ietf.org/id/draft-irtf-cfrg-xchacha-01.html). But as you can see, this RFC draft has already expired since January 24, 2020.

It is not about safe or not safe. This algorithm is only a customized version of chacha20-ietf-poly1305 in libsodium. Could you provide a strong reason about why we should support an algorithm that is not supported by any other libraries, and not accepted into RFC?

cc @LuoZijun

**Comment by zonyitoo** [comment]
shadowsocks-rust v1.10.6 have included this non-standard cipher.

**Comment by dev4u** [comment]
> shadowsocks-rust v1.10.6 have included this non-standard cipher.
> 
> Consider enable it by adding `--features "aead-cipher-extra"`.

这个需要在安卓app端做相应修改才行的。

**Comment by gengen1988** [comment]
I think `"aead-cipher-extra"` needs to be place here:

https://github.com/shadowsocks/shadowsocks-android/blob/44dc47fead4c866b6f167bbdcc88517a918d71af/core/build.gradle.kts#L41-L45
